addons/html_editor/static/tests/history.test.js

```diff
@@ -597,3 +566,84 @@ describe("destroy", () => {
         expect.verifySteps([]);
     });
 });
+
+describe("handleNewRecords", () => {
+    test("use handleNewRecords resource", async () => {
+        patchWithCleanup(Editor.prototype, {
+            dispatch(cmd, ...args) {
+                if (cmd === "CONTENT_UPDATED") {
+                    expect.step("CONTENT_UPDATED");
+                }
+                return super.dispatch(cmd, ...args);
+            },
+        });
+        const onChange = () => {
+            expect.step("onchange");
+        };
+        const resources = {
+            handleNewRecords: () => {
+                expect.step("handleNewRecords");
+            },
+        };
+        const { editor } = await setupEditor(`<p>[]</p>`, {
+            config: { onChange, resources },
+        });
+        expect.verifySteps([]);
+        await insertText(editor, "a");
+        expect.verifySteps([
+            "handleNewRecords",
+            "CONTENT_UPDATED",
+            "handleNewRecords",
+            "CONTENT_UPDATED",
+            "onchange",
+        ]);
+    });
+
+    test("undo/redo mutation apply with handleNewRecords resource", async () => {
```

Comment:
I would add a comment that explain why we need to hook into normalize as the information is hard to infer.

```
        // When handling new records, handlers could themselves generate
        // mutations. Those mutations must be added in `currentStep`, otherwise
        // we will lose those mutation in some history manipulation. For
        // instance, if the next added step is reverted, the generated mutations
        // will be reverted.
        //
        // To test against that kind of bug, we hooks into normalize to signal
        // the next `handleNewRecords` to add a mutation (the adding of the
        // `handle` class).
        //
        // With the current implementation (at the time of writing this test),
        // this bug can only occurs if a normalized trigger the
        // `handleNewRecords` to generate mutation. The reason is that
        // `handleObserverRecords` is called twice in `addStep` so if any
        // mutations is generated by the first `handleNewRecords`, they would be
        // added to the step by the second call to `handleNewRecords`.
```